File: src/storage.rs
use anyhow::{Context, Result};
use std::collections::HashSet;
use std::fs::{self, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;

use crate::config;
use crate::models::{StrokeData, Workout};

/// Path to the workouts JSONL file.
fn workouts_path() -> Result<PathBuf> {
    Ok(config::data_dir()?.join("workouts.jsonl"))
}

/// Path to a workout's stroke data file.
fn strokes_path(workout_id: u64) -> Result<PathBuf> {
    Ok(config::data_dir()?.join("strokes").join(format!("{}.jsonl", workout_id)))
}

/// Read all workouts from the JSONL file.
/// Returns an empty vec if the file doesn't exist.
pub fn read_workouts() -> Result<Vec<Workout>> {
    let path = workouts_path()?;
    if !path.exists() {
        return Ok(Vec::new());
    }

    let file = fs::File::open(&path)
        .with_context(|| format!("failed to open {}", path.display()))?;
    let reader = BufReader::new(file);
    let mut workouts = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        let line = line.with_context(|| format!("failed to read line {}", i + 1))?;
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let workout: Workout = serde_json::from_str(line)
            .with_context(|| format!("failed to parse workout on line {}", i + 1))?;
        workouts.push(workout);
    }

    Ok(workouts)
}

/// Append new workouts to the JSONL file, skipping any with IDs already present.
/// Returns the number of workouts actually written.
pub fn append_workouts(new_workouts: &[Workout]) -> Result<usize> {
    let existing = read_workouts()?;
    let existing_ids: HashSet<u64> = existing.iter().map(|w| w.id).collect();

    let path = workouts_path()?;
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&path)
        .with_context(|| format!("failed to open {} for appending", path.display()))?;

    let mut count = 0;
    for workout in new_workouts {
        if existing_ids.contains(&workout.id) {
            continue;
        }
        let json = serde_json::to_string(workout).context("failed to serialize workout")?;
        writeln!(file, "{}", json).context("failed to write workout")?;
        count += 1;
    }

    Ok(count)
}

/// Check if stroke data exists for a given workout ID.
pub fn has_stroke_data(workout_id: u64) -> Result<bool> {
    Ok(strokes_path(workout_id)?.exists())
}

/// Write stroke data for a workout.
pub fn write_stroke_data(workout_id: u64, strokes: &[StrokeData]) -> Result<()> {
    let path = strokes_path(workout_id)?;
    let mut file = OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .open(&path)
        .with_context(|| format!("failed to open {}", path.display()))?;

    for stroke in strokes {
        let json = serde_json::to_string(stroke).context("failed to serialize stroke")?;
        writeln!(file, "{}", json).context("failed to write stroke")?;
    }

    Ok(())
}

/// Read stroke data for a workout.
pub fn read_stroke_data(workout_id: u64) -> Result<Vec<StrokeData>> {
    let path = strokes_path(workout_id)?;
    if !path.exists() {
        return Ok(Vec::new());
    }

    let file = fs::File::open(&path)
        .with_context(|| format!("failed to open {}", path.display()))?;
    let reader = BufReader::new(file);
    let mut strokes = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        let line = line.with_context(|| format!("failed to read line {}", i + 1))?;
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let stroke: StrokeData = serde_json::from_str(line)
            .with_context(|| format!("failed to parse stroke on line {}", i + 1))?;
        strokes.push(stroke);
    }

    Ok(strokes)
}

/// Get the total count of stored workouts (without parsing them all).
pub fn workout_count() -> Result<usize> {
    let path = workouts_path()?;
    if !path.exists() {
        return Ok(0);
    }
    let file = fs::File::open(&path)?;
    let reader = BufReader::new(file);
    Ok(reader.lines().filter(|l| l.as_ref().is_ok_and(|s| !s.trim().is_empty())).count())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::HeartRate;

    fn test_workout(id: u64) -> Workout {
        Workout {
            id,
            user_id: 1,
            date: "2026-03-02 17:41:00".to_string(),
            timezone: Some("America/Denver".to_string()),
            distance: 5500,
            machine_type: "rower".to_string(),
            time: 19122,
            time_formatted: "31:52.2".to_string(),
            workout_type: Some("JustRow".to_string()),
            weight_class: Some("H".to_string()),
            stroke_rate: Some(26),
            stroke_count: Some(832),
            calories_total: Some(280),
            drag_factor: Some(83),
            heart_rate: Some(HeartRate {
                average: Some(118),
                min: None,
                max: None,
            }),
            pace_500m: Some("2:53.8".to_string()),
            average_watts: Some(67),
            comments: None,
            updated_at: None,
        }
    }

    #[test]
    fn test_workout_serialization() {
        let w = test_workout(1);
        let json = serde_json::to_string(&w).unwrap();
        let parsed: Workout = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.id, 1);
        assert_eq!(parsed.distance, 5500);
    }
}

File: src/storage.ts
import { appendFile, readFile, stat, writeFile } from "node:fs/promises";
import type { StrokeData, Workout } from "./models.ts";
import type { DataPaths } from "./paths.ts";

export interface StoreMeta {
  schema_version: number;
  created: string;
  last_sync?: string;
}

export const SCHEMA_VERSION = 1;

export async function readWorkouts(paths: DataPaths): Promise<Workout[]> {
  try {
    const text = await readFile(paths.workouts, "utf-8");
    return text
      .split("\n")
      .filter((line) => line.trim() !== "")
      .map((line) => JSON.parse(line) as Workout);
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return [];
    throw err;
  }
}

export async function appendWorkouts(paths: DataPaths, newWorkouts: Workout[]): Promise<number> {
  const existing = await readWorkouts(paths);
  const seen = new Set(existing.map((w) => w.id));

  const toWrite = newWorkouts.filter((w) => !seen.has(w.id));
  if (toWrite.length === 0) return 0;

  const lines = `${toWrite.map((w) => JSON.stringify(w)).join("\n")}\n`;
  await appendFile(paths.workouts, lines, "utf-8");
  return toWrite.length;
}

export async function workoutCount(paths: DataPaths): Promise<number> {
  try {
    const text = await readFile(paths.workouts, "utf-8");
    return text.split("\n").filter((line) => line.trim() !== "").length;
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return 0;
    throw err;
  }
}

export async function hasStrokeData(paths: DataPaths, workoutId: number): Promise<boolean> {
  try {
    await stat(paths.strokeFile(workoutId));
    return true;
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return false;
    throw err;
  }
}

export async function writeStrokeData(
  paths: DataPaths,
  workoutId: number,
  strokes: StrokeData[],
): Promise<void> {
  const lines = `${strokes.map((s) => JSON.stringify(s)).join("\n")}\n`;
  await writeFile(paths.strokeFile(workoutId), lines, "utf-8");
}

export async function readStrokeData(paths: DataPaths, workoutId: number): Promise<StrokeData[]> {
  let text: string;
  try {
    text = await readFile(paths.strokeFile(workoutId), "utf-8");
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return [];
    throw err;
  }
  const strokes: StrokeData[] = [];
  for (const line of text.split("\n")) {
    if (line.trim() === "") continue;
    try {
      const parsed = JSON.parse(line) as unknown;
      if (parsed != null && typeof parsed === "object" && !Array.isArray(parsed)) {
        strokes.push(parsed as StrokeData);
      }
    } catch {}
  }
  return strokes;
}

export async function readMeta(paths: DataPaths): Promise<StoreMeta | null> {
  let text: string;
  try {
    text = await readFile(paths.meta, "utf-8");
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code !== "ENOENT" && code !== "ENOTDIR") {
      console.error(`Warning: ${paths.meta} is unreadable and will be ignored.`);
    }
    return null;
  }
  try {
    return JSON.parse(text) as StoreMeta;
  } catch {
    console.error(`Warning: ${paths.meta} is corrupt and will be ignored.`);
    return null;
  }
}

export async function writeMeta(paths: DataPaths, meta: StoreMeta): Promise<void> {
  await writeFile(paths.meta, `${JSON.stringify(meta, null, 2)}\n`, "utf-8");
}
